### PR TITLE
Backport Bitcoin PR#8708: net: have CConnman handle message sending

### PR DESCRIPTION
**Pull request by OlegGirko**
This is backport of Bitcoin PR bitcoin/bitcoin#8708.

This PR moves responsibility for sending messages from `CNode` to `CConnman`.

The naive (textual) application of Bitcoin PR bitcoin/bitcoin#8708 caused a problem with Dash-specific code.

This PR introduces `CNode::nSendVersion` member that is initialised to 0 initially and then set to minimum of version received in `verack` message and `PROTOCOL_VERSION` when initial handshake is complete (`verack` message received after sending `version` message). `CConnman::PushMessage()` method requires `nSendVersion` to be non-zero and fails with assertion failure otherwise. In order to send message during handshake, `CConnman::PushMessageWithVersion()` method is used, but it's used only for initial `version`, `verack` messaged as well as `reject` messages in case initial handshake fails. This is OK because no other messages are accepted before initial handshake is completed.

However, Dash sends `getsporks` message in another thread without waiting for initial handshake to complete. As a temporary solution to avoid assertion failure caused by zero `nSendVersion`, I had to use `CConnman::PushMessageWithVersion()` method with `INIT_PROTO_VERSION` version for this message as well. However, I think, more correct solution would be better to send `getsporks` message only after initial handshake is completed.

The original PR description follows.
---
See individual commits for more description. In particular, bitcoin/bitcoin@75ead75.

This (along with bitcoin/bitcoin#8707) fixes the current maxupload test failure.

### CODE REVIEW

**Comment by UdjinM6** [comment]
I think this workaround for `getsporks` violates the idea of "handshake first" approach and should be reverted. Should probably modify `CopyNodeVector()` instead to have an optional param which would allow to return only nodes that completed handshake already (i.e. with `nSendVersion != 0`) and then call it and use results of that version in the main loop of `CMasternodeSync::ProcessTick()`.

File: src/main.cpp [hunk 1]
```diff
@@ -318,11 +319,36 @@ void UpdatePreferredDownload(CNode* node, CNodeState* state)
...
-    state.address = pnode->addr;
+void PushNodeVersion(CNode *pnode, CConnman& connman, int64_t nTime)
+{
+    ServiceFlags nLocalNodeServices = pnode->GetLocalServices();
+    uint64_t nonce = pnode->GetLocalNonce();
+    int nNodeStartingHeight = pnode->GetMyStartingHeight();
+    NodeId nodeid = pnode->GetId();
+    CAddress addr = pnode->addr;
+
+    CAddress addrYou = (addr.IsRoutable() && !IsProxy(addr) ? addr : CAddress(CService("0.0.0.0", 0), addr.nServices));
```
Comment (UdjinM6) [inline src/main.cpp]: It looks like we are missing https://github.com/bitcoin/bitcoin/pull/8128 and https://github.com/bitcoin/bitcoin/pull/8740 which were originally merged before 8708.
Were they skipped intentionally?

Comment (OlegGirko) [inline src/main.cpp]: They were skipped because there were too many PRs already. But I can backport them as well if needed.

**Comment by OlegGirko** [comment]
No, this workaround does not violate "handshake first" approach. Original design does. Dash was sending `getsporks` message from a separate thread without waiting for handshake completion before thid PR. This PR changes nothing in this regard, it just uses `CConnman::PushMessageWithVersion()` method instead of `CConnman::PushMessage()` to not trigger assertion failure. Before it, `CNode::PushMessage()` did not have this assert and was used in all cases.

**Comment by UdjinM6** [comment]
Well, yes, ok, that's the way to put it correctly :)
I guess we should fix it in a separate PR then.

utACK

**Comment by OlegGirko** [comment]
I completely agree. Let's hope that there's nothing in code that depends on this "feature".

File: src/masternode-sync.cpp [hunk 2]
```diff
@@ -324,12 +324,12 @@ void CMasternodeSync::ProcessTick()
         if(Params().NetworkIDString() == CBaseChainParams::REGTEST)
         {
             if(nRequestedMasternodeAttempt <= 2) {
-                pnode->PushMessage(NetMsgType::GETSPORKS); //get current network sporks
+                g_connman->PushMessageWithVersion(pnode, INIT_PROTO_VERSION, NetMsgType::GETSPORKS); //get current network sporks
```
Comment (tgflynn) [inline src/masternode-sync.cpp]: Do we know that this always runs before the peer handshake has been done ?

I would think that the order is actually indeterminate here and it seems re-sending the version could get us disconnected.

File: src/main.cpp [hunk 3]
```diff
@@ -319,10 +319,10 @@ void UpdatePreferredDownload(CNode* node, CNodeState* state)
 }
 
 void InitializeNode(NodeId nodeid, const CNode *pnode) {
+    CAddress addr = pnode->addr;
+    std::string addrName = pnode->addrName;
     LOCK(cs_main);
-    CNodeState &state = mapNodeState.insert(std::make_pair(nodeid, CNodeState())).first->second;
-    state.name = pnode->addrName;
-    state.address = pnode->addr;
+    mapNodeState.emplace_hint(mapNodeState.end(), std::piecewise_construct, std::forward_as_tuple(nodeid), std::forward_as_tuple(addr, std::move(addrName)));
```
Comment (tgflynn) [inline src/main.cpp]: IMO this change is an excessive optimization. The code is much less readable than the previous version and I doubt this code is a performance bottleneck. But I am willing to accept it to minimize divergence from upstream.

File: src/masternode-sync.cpp (hunk 2 above)
Comment (UdjinM6) [inline src/masternode-sync.cpp]: `PushMessageWithVersion` does not send `somemessage` and `version` messages, it sends `somemessage` serializing it for `someversion`. I think this shouldn't cause any issues.

Re checking `nVersion`: yes, we can check either `nVersion` or `nSendVersion` to make sure we received `version` message. But as it was said above it's not smth that this PR breaks or trying to fix, it's already how things work in current design. It was simply discovered here that things do not work as they should, so fixing this should probably be done in a separate PR that fixes exactly this issue.

**Comment by tgflynn** [comment]
utACK

**Pull request merged**